Feature: Earn by QuestionPro Survey Token Option

yongkanm opened the pull request:
### Description

A new type of token option (earn tokens by taking QuestionPro survey) is implemented in this pull request. This token option type allows a student to earn tokens by taking a QuestionPro survey. In addition to default configurations for a token option, the following configurations are available for this type of token option:

1.  QuestionPro Survey (single selection with options being surveys owned by the QuestionPro user whose credential is provided to Token ATM)
2.  Collect Email From (“Custom Variable” or “Student Response”)
    1.  If “Custom Variable” is selected, Custom Variable Name (text) needs to be configured.
    2.  If “Student Response” is selected, Question for Email (single selection with options of questions in the specified QuestionPro survey) needs to be configured.  
          
        Note: only the following three types of questions will be displayed:
        *   Contact Information: The email entry will be treated as a question
        *   Single Row Text: Each row will be treated as a separate question
        *   Email Address: Each row will be treated as a separate question
3.  Can Request From
4.  Can Request Until

Due to the QuestionPro API call limitation, the following mitigation is implemented:

*   For any configuration that involves communication with QuestionPro, the communication will be delayed until the user manually triggers it. Specifically, there are two configurations fall into this type: QuestionPro Survey and Question for Email. Both of them are single-selection fields, and their options will not be loaded until the user clicks the refresh button.
*   Field validations that requires communication with QuestionPro will be delayed until the token option is being saved. Specifically, after all other validations have been performed, the following two checks will be made:
    *   If the user did not load options for QuestionPro Survey field, the existence of the specified QuestionPro survey will be checked.
    *   If the user selected “Student Response” for the Collect Email From field and did not load options for the Question for Email field, the existence of the specified question will be checked.

A student's request will be approved only if all following conditions met:

1.  The request is submitted at a time no earlier than the start date.
2.  The request is submitted at a time no later than the end date.
3.  The student does not have an approved request for this token option.
4.  The student's email address in Canvas is found in the configured field (either a custom variable or a student's manual response to a question) of a completed survey response.

When approved, Token ATM will apply the configured token balance change.

### Provide QuestionPro Credential

This token option requires communication with QuestionPro, therefore QuestionPro credential needs to be provided. The following three fields are needed:

1.  [Environment (ENV)](https://www.questionpro.com/api/environments.html)
2.  User ID: could be obtained by clicking the avatar on the top-right corner → clicking My Account. Customer ID shown in this page is the User ID.
3.  API Key: could be obtained by clicking any survey → selecting Integration in the menu → clicking API. The API Key is obtainable by activate the API access. The user needs to be allowed by the organization manager to get API access.

Similar to #103, QuestionPro credential is optional, and the behaviors of viewing/editing token option and request processing are the same as described in #103 when QuestionPro credential is missing.

Note: due to the QuestionPro API call limitation, validation of QuestionPro credential will not be performed when saving it or logging in. Instead, a “Validate” button is added to the credential configuration modal for QuestionPro to allow the user to manually validate the provided QuestionPro credential.

### Testing Note

Please test if the new type of token option works as described.

yongkanm committed:
Implement Earn by QuestionPro Survey Token Option

yongkanm committed:
Enhance Question Resolution & Various Bug Fixes

yongkanm committed:
Exclude Incomplete Survey Response

yongkanm commented:
> **Note**: Configuring credential for QuestionPro environment value can be confusing.
> 
> I assume `Environment (ENV)` is meant to be `Environment Value` as stated in this doc: https://www.questionpro.com/api/environments.html The user focused documentation/UI on QuestionPro focuses on data center over environment value: https://www.questionpro.com/help/data-center.html
> 
> A dropdown/selection box with the options labeled would probably be a more user friendly approach. But we still need to decide if we put the focus on the data center or environment value in our UI. This also means Token ATM will be less flexible to changes made by Question Pro.
> 
> If we restrict the selection a decent format could be: `{{data center}} ({{ENV}})`, i.e. `US (com)` or `US (questionpro.com)`

I think provide both documentation link to the user in the Google Docs opened via "Help" button could be helpful. Given that domain name is more likely to be somehow changed, I prefer to allow the user to enter the domain suffix instead of having a single-selection field.

yongkanm commented:
> **Note**: API keys for Canvas and Qualtrics are masked with dots, but this isn't the case for QuestionPro.
> 
> Default masking is probably the better approach, but modern web UIs generally allow temporary unmasking based on user action (This would be a good feature to add in future).

Nice catch! I forget to mask the field of API key. I'm not sure whether allowing temporary unmasking will help too much since these secrets/API keys are usually copy-pasted.

yongkanm committed:
Upd-S ate QuestionPro Credential FormField

epsilonError reviewed (changes requested):
Functionality-wise the new Token Option works as expected.

But there are some edge cases for the data collected from Question Pro that need to be handled before this feature can be approved.

2 Errors: [Unhandled 404 error from no survey responses during Request Processing](https://github.com/UCI-IN4MATX-191-Token-ATM/token-atm-spa/pull/107#issuecomment-2041588735), and [QuestionPro Survey question names can include HTML tags which break the Token ATM configuration](https://github.com/UCI-IN4MATX-191-Token-ATM/token-atm-spa/pull/107#issuecomment-2041701529).

There are also some notes about areas that could use changes, but the **note**s are not blockers for this pull request. If they are quick to take care of please do, otherwise we'll tackle them in future.

yongkanm commented on src/app/services/question-pro.service.ts:
```diff
@@ -0,0 +1,241 @@
+        if (this.participationCache.get(surveyId)?.has(typeStr)) return;
+        this.participationCache.get(surveyId)?.set(typeStr, new Set<string>());
+        const tmpSet = this.participationCache.get(surveyId)?.get(typeStr);
+        const responses = await this.getResponses(surveyId);
```
Thank you for writing this wrapper! I'll apply it to `getQuestions()` and `getResponses()` to handle 404 errors specially.

yongkanm committed:
Various fixes

yongkanm commented:
> **Note**: Refreshing the `Question for Email` single selection for a Survey without any configured questions, displays a 404 `axiosError` inline in the validation error field.
> 
> This is the proper place for the error to be displayed, but displaying the user readable error provided by Question Pro would be an improvement. Something from the provided text could be a good start: `"https://www.questionpro.com/api/v2/#error-codes","The resource that youre trying to access doesnt exist. - Question not present","NOT_FOUND"`

> **Error**: Processing a Student Request for a Token Option for a Survey with no Responses results in an error.
> 
> A 404 error is returned by Question Pro. A survey with no responses should be treated as an empty set, handling this 404 error as an empty set should fix this problem.

Thanks for discovering these edge cases. I've applied the wrapper you provided to `getQuestions()` and `getResponses()`. For `getResponses()` an empty set of response will be returned in the case of 404 error. For `getQuestions()`, an error with more understandable error message is thrown.

yongkanm commented:
> > > A dropdown/selection box with the options labeled would probably be a more user friendly approach. But we still need to decide if we put the focus on the data center or environment value in our UI. This also means Token ATM will be less flexible to changes made by Question Pro.
> > > If we restrict the selection a decent format could be: `{{data center}} ({{ENV}})`, i.e. `US (com)` or `US (questionpro.com)`
> > 
> > 
> > I think provide both documentation link to the user in the Google Docs opened via "Help" button could be helpful. Given that domain name is more likely to be somehow changed, I prefer to allow the user to enter the domain suffix instead of having a single-selection field.
> 
> I've added both links to the 'Help' documentation and will add step-by-step instruction soon. Since configuring credentials should be an infrequent action, I think going with straight text input is a reasonable approach especially if combined with directions to follow.
> 
> > > I assume `Environment (ENV)` is meant to be `Environment Value` as stated in this doc: https://www.questionpro.com/api/environments.html The user focused documentation/UI on QuestionPro focuses on data center over environment value: https://www.questionpro.com/help/data-center.html
> 
> **Note**: The Credential Configuration UI says `Environment (ENV)` while the [reference table](https://www.questionpro.com/api/environments.html) calls this `Environment Value`, what do you think about having `Environment Value (ENV)` in the Token ATM UI instead?

Thank you for updating the help documentation for credentials! And I agree with you that making the field name exactly matches the documentation would be helpful, so I've updated the first field name to "Environment Value (ENV)."

yongkanm commented:
> **Note**: Axios seems to make requests even when credential details are not provided.
> 
> Even when a configuration is not provided, an http request is sent out to services. For Question Pro the returned error is `[{"code":"1","isAxiosError":true},"ENOTFOUND"]`.
> 
> The `rawAPIRequest`s, or external service calls using `AxiosService.request`, should preemptively check that credentials exist. It doesn't make sense to make HTTP requests when we know the necessary info hasn't been provided.

Thanks for pointing this out! I did not add a check before since unless a valid credential is provided, no API call cost will occur. But you are right that it would be better to check the credential locally before making a request to a remote server. I've added such check to `#rawAPIRequest()` and `#paginatedRequest()`.

yongkanm commented:
> **Error**: Question Pro survey question names can break Token ATM Configuration.
> 
> Question Pro allows Survey Question Names to be styled with HTML tags. And including HTML in the Token ATM configuration will cause it to not parse correctly.
> 
> To Reproduce:
> 
> 1. Add a Single Row question to a Question Pro survey and add a line break as the first character of the name by pressing enter.
> 2. When selecting that question with the Token ATM single selection field you'll see the name as `<br />Name`.
> 3. Save this Token Option.
> 4. Open the Switch Course menu, and select the same course.
> 5. Error Occurs, the make a new Token ATM configuration modal appears instead of loading the course.
> 
> It's possible to style the names with a Rich Text editor or a Source Code editor (and also remove the styling) on Question Pro. The raw source is what is displayed as the selectable question name.
> 
> We may want to strip all the html for the displayed names in the Token ATM selection fields. And safely encode the source before updating the Token ATM configuration.

Thank you for discovering this issue! I've tested and confirmed that this issue also applies to survey name. To resolve it, I store both survey name and the survey question name (if applicable) in the base64 form. Please note that **this is a breaking change**, so please delete all existing token options with the type of earn-by-question-pro-survey before pulling the latest commit.

I currently did not implement stripping of HTML from options in the selection field since it does not influence the functionality (`ngx-mat-select-search` will display the raw string instead of interpreting HTML). I think we can implement it in a future pull request if the user found it confusing.

yongkanm commented:
> **Note**: When setting up multiple QuestionPro Token Options, each new modal window requires an API call to refresh the Survey List.
> 
> Would it be better to cache this for the browser session? User's can hit the refresh button at any time, if they have made changes since the last caching.
> 
> The same idea also covers the Survey Questions single selection field, but I expect Survey Questions to be updated more frequently than Surveys and maintaining the second-order mapping of all questions for each survey would be a lot.
> 
> **P.S.**: This would also complicate the work around we may need to implement for the Question Pro API not having easy access to the Surveys shared with the current user. (Which would additionally require resolving an inverse mapping from Survey ID to the proper API token)

That makes sense, but it will introduce extra complexity when loading options for selection fields as you mentioned. Specifically, determining whether cached data should be treated as validated data when saving a token option could be difficult. If we treat cached data as validated data, then there is a possibility of saving a token option with a survey that has already been deleted. If we still perform validation for it, then the cached data lost its meaning since an API request will be made anyway. I think it would be better to be considered a future enhancement when needed.

epsilonError reviewed (approved):
https://github.com/UCI-IN4MATX-191-Token-ATM/token-atm-spa/pull/107#pullrequestreview-1985353342

Errors have been addressed. And all tests still pass. Pull request Approved.
